lib/unlock-helpers.js

```diff
@@ -0,0 +1,222 @@
+};
+
+helpers.encodePassword = function (key) {
+  return key.replace(/\s/ig, "%s");
+};
+
+helpers.stringKeyToArr = function (key) {
+  return key.trim().replace(/\s+/g, '').split(/\s*/);
+};
+
+helpers.pinUnlock = async function (adb, driver, capabilities) {
+  logger.info(`Trying to unlock device using pin ${capabilities.unlockKey}`);
+  await helpers.dismissKeyguard(driver, adb);
+  let keys = helpers.stringKeyToArr(capabilities.unlockKey);
+  if (await adb.getApiLevel() >= 21) {
+    let els = await driver.findElOrEls("id", "com.android.systemui:id/digit_text", true);
+    if (_.isEmpty(els)) {
+      throw new Error("Error finding unlock pin buttons!");
+    }
+    let pins = {};
+    for (let e of els) {
+      let text = await driver.getAttribute("text", e.ELEMENT);
+      pins[text] = e;
+    }
+    for (let pin of keys) {
+      let el = pins[pin];
+      await driver.click(el.ELEMENT);
+    }
+    let el = await driver.findElOrEls("id", "com.android.systemui:id/key_enter", false);
+    await driver.click(el.ELEMENT);
+  } else {
+    for (let pin of keys) {
+      let el = await driver.findElOrEls("id", `com.android.keyguard:id/key${pin}`, false);
+      if (el === null) {
+        throw new Error(`Error finding unlock pin '${pin}' button!`);
+      }
+      await driver.click(el.ELEMENT);
+    }
+    let el = await driver.findElOrEls("id", "com.android.keyguard:id/key_enter", false);
+    await driver.click(el.ELEMENT);
+  }
+  // Waits a bit for the device to be unlocked
+  await sleep(UNLOCK_WAIT_TIME);
+};
+
+helpers.passwordUnlock = async function (adb, driver, capabilities) {
+  logger.info(`Trying to unlock device using password ${capabilities.unlockKey}`);
+  await helpers.dismissKeyguard(driver, adb);
+  let key = capabilities.unlockKey;
+  // Replace blank spaces with %s
+  key = helpers.encodePassword(key);
+  // Why adb ? It was less flaky
+  await adb.shell(["input", "text", key]);
+  // Why sleeps ? Avoid some flakyness waiting for the input to receive the keys
+  await sleep(INPUT_KEYS_WAIT_TIME);
+  await adb.shell(["input", "keyevent", KEYCODE_NUMPAD_ENTER]);
+  // Waits a bit for the device to be unlocked
+  await sleep(UNLOCK_WAIT_TIME);
+};
+
+helpers.getPatternKeyPosition = function (key, initPos, piece) {
+  /*
+  How the math works:
+  We have 9 buttons divided in 3 columns and 3 rows inside the lockPatternView,
+  every button has a position on the screen corresponding to the lockPatternView since
+  it is the parent view right at the middle of each column or row.
+  */
+  const cols = 3;
+  const pins = 9;
+  let xPos = (key, x, piece) => {
+    return Math.round(x + ((key % cols) || cols) * piece - piece / 2);
+  };
+  let yPos = (key, y, piece) => {
+    return Math.round(y + (Math.ceil(((key % pins) || pins) / cols) * piece - piece / 2));
+  };
+  return {x: xPos(key, initPos.x, piece), y: yPos(key, initPos.y, piece)};
+};
+
+helpers.getPatternActions = function (keys, initPos, piece) {
```

Comment:
I think there should be a verification of the fact that keys array should at least include 2 elements.